scrapy/crawler.py

```diff
@@ -72,42 +72,55 @@ def __init__(
         self.spidercls: Type[Spider] = spidercls
         self.settings: Settings = settings.copy()
         self.spidercls.update_settings(self.settings)
+        self._update_root_log_handler()
 
         self.addons: AddonManager = AddonManager(self)
-        self.addons.load_settings(self.settings)
-
         self.signals: SignalManager = SignalManager(self)
 
-        self.stats: StatsCollector = load_object(self.settings["STATS_CLASS"])(self)
-
-        handler = LogCounterHandler(self, level=self.settings.get("LOG_LEVEL"))
-        logging.root.addHandler(handler)
+        self._init_reactor: bool = init_reactor
+        self.crawling: bool = False
+        self._settings_loaded: bool = False
+        self._started: bool = False
 
-        d = dict(overridden_settings(self.settings))
-        logger.info(
-            "Overridden settings:\n%(settings)s", {"settings": pprint.pformat(d)}
-        )
+        self.extensions: Optional[ExtensionManager] = None
+        self.stats: Optional[StatsCollector] = None
+        self.logformatter: Optional[LogFormatter] = None
+        self.request_fingerprinter: Optional[RequestFingerprinter] = None
+        self.spider: Optional[Spider] = None
+        self.engine: Optional[ExecutionEngine] = None
 
+    def _update_root_log_handler(self) -> None:
         if get_scrapy_root_handler() is not None:
             # scrapy root handler already installed: update it with new settings
             install_scrapy_root_handler(self.settings)
+
+    def _apply_settings(self) -> None:
+        if self._settings_loaded:
```

Comment:
Can it be simplified/improved to `if self.settings.frozen`?

Reply:
Also, if it's called when the settings are frozen, should it be a warning?

Reply:
Or an error?

Reply:
Right, and error is even better, if there is no use case for running this method twice

Reply:
Though when designing it I thought that we need a check so that we can safely run it twice, e.g. directly and then via crawl(). This only matters in tests as users shouldn't run _apply_settings() directly, and either a warning or an error will affect them, as we indeed have tests that call `get_crawler()` (which calls _apply_settings() so that you can get an initialized crawler) and then call `crawl()`.